Libraries/MLXVLM/Models/Paligemma.swift

```diff
@@ -0,0 +1,743 @@
+        PaliGemmaMultiModalProjector
+
+    public let config: PaliGemmaConfiguration
+
+    public var vocabularySize: Int { config.vocabularySize }
+    public var kvHeads: [Int] { languageModel.kvHeads }
+    public var headDim: MLX.IntOrPair { languageModel.headDim }
+
+    public func loraLinearLayers() -> MLXLMCommon.LoRALinearLayers {
+        languageModel.model.layers.map { ($0.attention, ["q_proj", "v_proj"]) }
+    }
+
+    public init(_ config: PaliGemmaConfiguration) {
+        self.config = config
+        self._visionModel.wrappedValue = Vision.VisionModel(config.visionConfiguration)
+        self._languageModel.wrappedValue = Language.LanguageModel(config.textConfiguration)
+        self._multiModalProjector.wrappedValue = PaliGemmaMultiModalProjector(
+            config.visionConfiguration)
+    }
+
+    private func inputEmbeddings(inputIds: MLXArray, pixelValues: MLXArray?, mask: MLXArray) -> (
+        MLXArray, MLXArray
+    ) {
+        guard let pixelValues else {
+            return (inputIds, mask)
+        }
+
+        let inputEmbedding = languageModel.model.embedTokens(inputIds)
+        let (hiddenState, _, _) = self.visionModel(
+            pixelValues.transposed(0, 2, 3, 1).asType(inputEmbedding.dtype),
+            outputHiddenStates: true
+        )
+
+        var imageFeatures = hiddenState[.newAxis, .ellipsis].asType(pixelValues.dtype)
+        imageFeatures = multiModalProjector(imageFeatures)
+
+        return prepareInputsForMultimodal(
+            imageFeatures: imageFeatures, inputEmbedding: inputEmbedding,
+            inputIds: inputIds, attentionMask: mask)
+    }
+
+    private func prepareInputsForMultimodal(
+        imageFeatures: MLXArray, inputEmbedding: MLXArray, inputIds: MLXArray,
+        attentionMask: MLXArray
+    ) -> (MLXArray, MLXArray) {
+        let embedDimension = imageFeatures.dim(2)
+        let (batchSize, sequenceLength) = inputIds.shape2
+        var scaledImageFeatures = imageFeatures / pow(Float(config.hiddenSize), 0.5)
+        var finalEmbedding = zeros([batchSize, sequenceLength, embedDimension])
+
+        let textMask = (inputIds .!= config.imageTokenIndex) & (inputIds .!= config.padTokenId)
+        let imageMask = inputIds .== config.imageTokenIndex
+        let padMask = inputIds .== config.padTokenId
+
+        // expand masks to match embedding dimension
+        var textMaskExpanded = expandedDimensions(textMask, axis: -1)
+        textMaskExpanded = repeated(textMaskExpanded, count: embedDimension, axis: -1)
+        var padMaskExpanded = expandedDimensions(padMask, axis: -1)
+        padMaskExpanded = repeated(padMaskExpanded, count: embedDimension, axis: -1)
+
+        // insert padding and text token embeddings
+        finalEmbedding = which(textMaskExpanded, inputEmbedding, finalEmbedding)
+        finalEmbedding = which(padMaskExpanded, zeros(like: finalEmbedding), finalEmbedding)
```

Comment:
This `zeros` also should be a plain scalar and inherit the type of the `padMaskExpanded`.